sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -278,77 +280,185 @@ class RocksDB(
+      (Long, Option[String]) = {
+    lineage.foreach {
+      case LineageItem(version, uniqueId) =>
+        if (latestSnapshotVersionsAndUniqueIds.contains((version, Some(uniqueId)))) {
+          return (version, Some(uniqueId))
+        }
+    }
+    throw QueryExecutionErrors.cannotGetLatestSnapshotVersionAndUniqueIdFromLineage(
+      printLineageItems(lineage), printLineage(latestSnapshotVersionsAndUniqueIds)
+    )
+  }
+
+  /**
+   * Read the lineage from the changelog files. It first get the changelog reader
+   * of the correct changelog version and then read the lineage information from the file.
+   * The changelog file is named as version_stateStoreCkptId.changelog
+   * @param version version of the changelog file, used to load changelog file.
+   * @param stateStoreCkptId uniqueId of the changelog file, used to load changelog file.
+   * @return
+   */
+  private def getLineageFromChangelogFile(
+      version: Long,
+      stateStoreCkptId: Option[String]): Array[LineageItem] = {
+    var changelogReader: StateStoreChangelogReader = null
+    var currLineage: Array[LineageItem] = Array.empty
+    try {
+      changelogReader = fileManager.getChangelogReader(version, stateStoreCkptId)
+      currLineage = changelogReader.lineage
+      logInfo(log"Loading lineage: " +
+        log"${MDC(LogKeys.LINEAGE, lineageManager)} from " +
+        log"changelog version: ${MDC(LogKeys.VERSION_NUM, version)} " +
+        log"uniqueId: ${MDC(LogKeys.UUID, stateStoreCkptId.getOrElse(""))}.")
+    } finally {
+      if (changelogReader != null) {
+        changelogReader.closeIfNeeded()
+      }
+    }
+    currLineage
+  }
+
+
   /**
    * Load the given version of data in a native RocksDB instance.
    * Note that this will copy all the necessary file from DFS to local disk as needed,
    * and possibly restart the native RocksDB instance.
    */
-  def load(
+  private def loadV2(
+        version: Long,
+        stateStoreCkptId: Option[String],
+        readOnly: Boolean = false): RocksDB = {
+  // An array contains lineage information from [snapShotVersion, version] (inclusive both end)
+  var currVersionLineage: Array[LineageItem] = lineageManager.getLineage()
+  try {
+    if (loadedVersion != version || (loadedStateStoreCkptId.isEmpty ||
+        stateStoreCkptId.get != loadedStateStoreCkptId.get)) {
+      closeDB(ignoreException = false)
+
+      val (latestSnapshotVersion, latestSnapshotUniqueId) = {
+        // Special handling when version is 0.
+        // When loading the very first version (0), stateStoreCkptId does not need to be defined
+        // because there won't be 0.changelog / 0.zip file created in RocksDB.
+        if (version == 0 && stateStoreCkptId.isEmpty) {
+          (0L, None)
+        // When there is a snapshot file, it is the ground truth, we can skip
+        // reconstructing the lineage from changelog file.
+        } else if (fileManager.existsSnapshotFile(version, stateStoreCkptId)) {
```

Comment:
The check from local lineage manager already happens here above: https://github.com/apache/spark/blob/3245b6275d2c55ab948dcf88c99525d6120fdaa0/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala#L348

When version != loadedVersion, it means the values in memory are not valid anymore and we need to reload from dfs. So I believe check local lineage manager is not a valid option here.

Reply:
I think you are suggesting that when `loadedVersion > version`, we should not reload from the disk and trust everything from the memory (including the lineage and all other variables including `latestSnapshotVersions`, etc). 

I suggest we do not do this in this PR because:

This also applies to V1, because lineage is the only new information added in V2. So now we are touching a long-standing logic for performance improvement to all rocksDB load() function. The scope now is larger than this PR because this PR should focus only on V2 integration first. In case the change caused some failures in loading, (because we now trust information from memory, which is more error-prone than reloading from disk), we can only revert the performance improvement PR, not this one.